文書検索層（想定質問精度検証用）

（Question2Pre_Question.ipynb の続き）

## openaiに関する設定

In [2]:
!pip install openai==1.7

  Attempting uninstall: typing-extensions
    Found existing installation: typing_extensions 4.5.0
    Uninstalling typing_extensions-4.5.0:
      Successfully uninstalled typing_extensions-4.5.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.
llmx 0.0.15a0 requires tiktoken, which is not installed.
tensorflow-probability 0.22.0 requires typing-extensions<4.6.0, but you have typing-extensions 4.9.0 which is incompatible.


In [14]:
import os
import openai
# OpenAIに関する環境変数
api_key = os.environ["OPENAI_API_KEY"]

In [17]:
from openai import OpenAI
import pandas as pd

# OpenAI クライアントの初期化
client = OpenAI(api_key=api_key)

## 参照データの導入

In [7]:

# データフレームの取得
file_path = 'C:/Users/iput Hata Sakuto/cuda-11.8/提出/自作モデル-2/0. 想定質問作成/q_embedding_df.csv'
base_df = pd.read_csv(file_path, encoding='utf-8')
base_df

chunk  \
0   現代社会の課題は多様な要素が複雑に絡み合い、しかも常に変化しています。その解決には、単一の専...   
1   55年ぶりに国がつくった新しい大学制度によって生まれた、職業に直結した教育を行う専門職大学。...   
2   東京国際工科専門職大学では、「プロフェッショナル」とは単に専門職に従事する人というだけではな...   
3   学びが行き届く 「少人数教育」原則40人以下で行う少人数授業が、専門職大学の特長。本学ではさ...   
4   実践を学べる「実務家教員」 教員の４割以上は企業等で実務を経験した「実務家教員」特に本学では...   
5   ビジネスの最前線「企業内実習」 一般的な大学で行われている短期の就業体験・インターンシップと...   
6   授業全体での「実習・実技の割合」理論と実践の双方を重視するため、卒業に必要な単位のうち、1/...   
7   ビジネスの現場で培う高度な実践力 産業界との連携が強く、ビジネスの実情に即して授業内容の改善...   
8   工科学部・情報工学科 AI戦略コース 人工知能の理論から、実装のためのプログラミング技術まで...   
9   １年次体験的な学びからスタートし、ＩＣＴの基本をマスター。― 実践的イントロダクション ― ...   
10  ２年次コース別のカリキュラムスタート　― 「エンジニアリングスキル」を身につける ― コース...   
11  ３年次専門知識・技能のレベルアップ　― 「テクノロジー実践力」と「ビジネスセンス」を磨く ―...   
12  ４年次応用力と総合力を備えた人材へ　― デジタル社会で広がる活躍の場 ― 「卒業研究制作」を...   
13  デザインエンジニアリング概論 “Designer in Society”の核となる、デザイン...   
14  ソリューション開発Ⅰ・Ⅱ 協力企業・団体から提供される実践的な課題をもとに、チームでソリュー...   
15  人工知能システム開発Ⅰ・Ⅱ 2年次前期、3年次前期の実習科目です。人工知能のシステム開発の専...   
16  深層学習 深層学習(Deep Learning) の主要技術であるニューラルネットワークの技...   

                       question  label  label_num  \
0                東京国際工科専門職大学の使命  専門職大学          0   
1                     専門職大学とは何か  専門職大学          0   
2   東京国際工科専門職大学の定義するプロフェッショナルとは  専門職大学          0   
3                       少人数教育とは     特徴          1   
4                        教員の特徴は     特徴          1   
5                       企業内実習とは     特徴          1   
6                        授業の特徴は     特徴          1   
7                        産学連携とは     特徴          1   
8                   AI戦略コースについて     特徴          1   
9                   1年生のとき学ぶことは     学び          2   
10                  2年生のとき学ぶことは     学び          2   
11                  3年生のとき学ぶことは     学び          2   
12                  4年生のとき学ぶことは     学び          2   
13             デザインエンジニアリング概論とは     授業          3   
14                  ソリューション開発とは     授業          3   
15                 人工知能システム開発とは     授業          3   
16                       深層学習とは     授業          3   

                                          q_embedding  \
0   [-0.00941627100110054, 0.005360241513699293, -...   
1   [0.01832834631204605, -0.015213198959827423, 0...   
2   [0.001960532506927848, -0.01159671787172556, -...   
3   [-0.016901545226573944, -0.010819817893207073,...   
4   [-0.0067285760305821896, 0.009487291797995567,...   
5   [-0.015425135381519794, -0.028941858559846878,...   
6   [-0.009149088524281979, 0.009476328268647194, ...   
7   [-0.02425614930689335, -0.005011129193007946, ...   
8   [-0.01949276775121689, -0.022640325129032135, ...   
9   [0.0070157707668840885, -0.008300903253257275,...   
10  [-0.004014338366687298, -0.00980047695338726, ...   
11  [4.047681068186648e-05, -0.011337393894791603,...   
12  [0.0010889001423493028, -0.011816248297691345,...   
13  [0.004818136338144541, -0.006898695603013039, ...   
14  [0.015440679155290127, -0.014237872324883938, ...   
15  [-0.003598037175834179, -0.01152889709919691, ...   
16  [-0.006485281977802515, -0.010722515173256397,...   

                                          a_embedding  
0   [-0.011608361266553402, 0.010893057100474834, ...  
1   [0.014174927957355976, -0.010714853182435036, ...  
2   [0.002627256792038679, -0.008873092010617256, ...  
3   [-0.006027283146977425, 0.010709190741181374, ...  
4   [-0.001338495989330113, -0.008723043836653233,...  
5   [-0.021340254694223404, -0.021042168140411377,...  
6   [0.0038443717639893293, 0.001330939820036292, ...  
7   [-0.014218892902135849, -0.026790393516421318,...  
8   [-0.0006897911080159247, -0.011758310720324516...  
9   [0.003032150911167264, -0.01046228688210249, 0...  
10  [0.0013306819600984454, -0.00452090660110116, ...  
11  [0.0027723119128495455, -0.0116502083837986, 0...  
12  [-0.01769779622554779, -0.011677035130560398, ...  
13  [0.0005100962007418275, 0.0037174364551901817,...  
14  [0.01210450753569603, -0.01519758440554142, 0....  
15  [0.007826392538845539, -0.014842928387224674, ...  
16  [-0.017577247694134712, 0.008625630289316177, ...

## 検証用データの導入

In [6]:
# データフレームの取得
test_file_path = 'C:/Users/iput Hata Sakuto/cuda-11.8/提出/自作モデル-2/0. 想定質問作成/test_question_Q.csv'
df_test = pd.read_csv(test_file_path, encoding='utf-8')
df_test

test_question  label_num  label  \
0                             東京国際工科専門職大学の使命について質問ですか？          0  専門職大学   
1                                 東京国際工科専門職大学の使命は何ですか？          0  専門職大学   
2                         東京国際工科専門職大学はどのような使命を持っていますか？          0  専門職大学   
3                           東京国際工科専門職大学の使命について教えてください。          0  専門職大学   
4                           東京国際工科専門職大学の使命について知りたいですか？          0  専門職大学   
..                                                 ...        ...    ...   
675  深層学習について詳しく説明してください。\n深層学習はどのような仕組みで動作しているのですか...          3     授業   
676                                 「深層学習はどのような技術ですか？」          3     授業   
677  深層学習とは何ですか？\n深層学習について教えてください。\n深層学習とはどのようなものです...          3     授業   
678                                 「深層学習の定義を教えてください。」          3     授業   
679                               深層学習はどのようなことを指すのですか？          3     授業   

                                           q_embedding  
0    [-0.0007721076835878193, -0.00861017033457756,...  
1    [0.0010231389896944165, -0.0049309502355754375...  
2    [0.004724525846540928, -0.009007260203361511, ...  
3    [-0.004020273219794035, -0.00677361199632287, ...  
4    [0.0009267496643587947, -0.00958362314850092, ...  
..                                                 ...  
675  [0.001292953733354807, 0.0005758868646807969, ...  
676  [0.0025534157175570726, -0.005754310637712479,...  
677  [-0.005072268657386303, -0.009966022334992886,...  
678  [-0.007288273889571428, -0.007133752107620239,...  
679  [-0.003823702922090888, -0.013265907764434814,...  

[680 rows x 4 columns]

## ベクトルの類似度比較による文書検索

### 想定質問の回答率

In [8]:
import ast
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np
from openai import OpenAI
import pandas as pd

In [9]:
#質問設定
pre_q = df_test["q_embedding"].apply(lambda x: np.array(ast.literal_eval(x)))

#元データの読み込み
base_df['q_embedding'] = base_df['q_embedding'].apply(lambda x: np.array(ast.literal_eval(x)))

pre_num = []

for i in range(len(pre_q)):
# コサイン類似度を計算
    similarities = cosine_similarity([pre_q[i]], list(base_df['q_embedding']))

    # 類似度が最大の行を取得
    most_similar_index = similarities.argmax()
    most_similar_chunk = base_df.loc[most_similar_index, 'chunk']
    most_similar_label = base_df.loc[most_similar_index, 'label_num']
    # 類似度も取得
    max_similarity = similarities[0, most_similar_index]

    # 結果の出力
    print(f"質問文[{i}]：\n",df_test['test_question'][i])
    print(f"最大類似度:: {max_similarity:.3f}")
    print("最も類似度が高い行のchunk:", most_similar_chunk)
    print("最も類似度が高い行のlabel:", most_similar_label)
    print("＝"*30)
    pre_num.append(most_similar_label)
print(pre_num)

質問文[0]：
 東京国際工科専門職大学の使命について質問ですか？
最大類似度:: 0.937
最も類似度が高い行のchunk: 現代社会の課題は多様な要素が複雑に絡み合い、しかも常に変化しています。その解決には、単一の専門分野の知識＜静的な専門知＞だけでは困難です。だからこそ私たちは、既存の諸学問の壁を超えて専門分野を横断し、現場の実践で得られた知見や最新テクノロジーを統合的に駆使した＜動的な創造知＞を目指します。そのために、学生と教員が一体となって探究し、産業界や地域、世界の人々と協働する。理論と実践を武器に新たな価値を創造し、より良い未来を創るために必要な力を共に育む。そして、多様で複雑な社会課題に立ち向かう“Designer in Society（社会とともにあるデザイナー）”を世に送り出す。これが、これからの時代に対応する新大学としての、私たち東京国際工科専門職大学の使命です。
最も類似度が高い行のlabel: 0
＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝
質問文[1]：
 東京国際工科専門職大学の使命は何ですか？
最大類似度:: 0.968
最も類似度が高い行のchunk: 現代社会の課題は多様な要素が複雑に絡み合い、しかも常に変化しています。その解決には、単一の専門分野の知識＜静的な専門知＞だけでは困難です。だからこそ私たちは、既存の諸学問の壁を超えて専門分野を横断し、現場の実践で得られた知見や最新テクノロジーを統合的に駆使した＜動的な創造知＞を目指します。そのために、学生と教員が一体となって探究し、産業界や地域、世界の人々と協働する。理論と実践を武器に新たな価値を創造し、より良い未来を創るために必要な力を共に育む。そして、多様で複雑な社会課題に立ち向かう“Designer in Society（社会とともにあるデザイナー）”を世に送り出す。これが、これからの時代に対応する新大学としての、私たち東京国際工科専門職大学の使命です。
最も類似度が高い行のlabel: 0
＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝＝
質問文[2]：
 東京国際工科専門職大学はどのような使命を持っていますか？
最大類似度:: 0.952
最も類似度が高い行のchunk: 現代社会の課題は多様な要素が複雑に絡み合い、しかも常に変化しています。

In [10]:
Ans = df_test["label_num"]
Ans = Ans.to_numpy()
Ans

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [11]:
data_len = len(Ans)

correct_data = 0
for i in range(data_len):
    if pre_num[i] == Ans[i]:
        correct_data += 1

print(correct_data/data_len)

0.9970588235294118


<結果>

q_embedding_df.csv の「question」をベースにテストケースにおける想定質問を作成したところ、約0.9971の正答率を得ることができた。

今回は事前に情報に対する想定質問を作成したことで上記の結果が得られた。このテストケースで有用性が見出せたと当方は考える。

---
以下、対象外

## 動作確認用

### 質問文を直に入力する場合のembedding

In [18]:
path = 'C:/Users/iput Hata Sakuto/cuda-11.8/提出/自作モデル-2/0. 想定質問作成/q_embedding_df.csv'
df = pd.read_csv(path, encoding='utf-8')

In [ ]:
def q_embedding(text, model="text-embedding-ada-002"):
    return client.embeddings.create(input=[text], model=model).data[0].embedding


q = input("質問をどうぞ：")
print("質問：",q)
q_emb = q_embedding(q, model='text-embedding-ada-002')
#print(q_emb)

### 最も類似度の高いものを出力する場合

In [ ]:
import ast
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

# データの読み込み
file_path = 'C:/Users/iput Hata Sakuto/cuda-11.8/Univercity/QA/q_embedding_df.csv'
embedded_df = pd.read_csv(file_path, encoding='utf-8') # embedded_df.csv を適切なファイルパスに置き換えてください
embedded_df['q_embedding'] = embedded_df['q_embedding'].apply(ast.literal_eval)

# コサイン類似度を計算
similarities = cosine_similarity([q_emb], list(embedded_df['q_embedding']))

# 類似度が最大の行を取得
most_similar_index = similarities.argmax()
most_similar_chunk = embedded_df.loc[most_similar_index, 'chunk']

# 類似度も取得
max_similarity = similarities[0, most_similar_index]

# 結果の出力
print(f"最大類似度:: {max_similarity:.3f}")
print("最も類似度が高い行のchunk:", most_similar_chunk)

最大類似度:: 0.873
最も類似度が高い行のchunk: デザインエンジニアリング概論 “Designer in Society”の核となる、デザインエンジニアリングの概念およびその方法論を学びます。実際の製品の分解作業を通じて、その機能や内部構造をハードウェアとソフトウェア両面から調査し、製品に求められる要素を分析します。デザインは製品だけでなくサービスも対象とし、製品やサービスの背後に存在する代表的な技術と設計のプロセスを理解することで、実際の制作に活かすデザインエンジニアリングの方法論を身につけます。


### 類似度の傾向を見る場合

In [ ]:
import ast
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

# データの読み込み
embedded_df = pd.read_csv('C:/Users/iput Hata Sakuto/cuda-11.8/Univercity/QA/q_embedding_df.csv',encoding='utf-8')  # embedded_df.csv を適切なファイルパスに置き換えてください
embedded_df['q_embedding'] = embedded_df['q_embedding'].apply(ast.literal_eval)

# コサイン類似度を計算
similarities = cosine_similarity([q_emb], list(embedded_df['q_embedding']))[0]

# 類似度の上位5個を取得
top_5_indices = np.argsort(similarities)[-5:][::-1]

# 上位5個の行と類似度を表示
for idx in top_5_indices:
    chunk = embedded_df.loc[idx, 'chunk']
    similarity = similarities[idx]
    print(f"類似度: {similarity:.3f}, \n {chunk}")

類似度: 0.873, 
 デザインエンジニアリング概論 “Designer in Society”の核となる、デザインエンジニアリングの概念およびその方法論を学びます。実際の製品の分解作業を通じて、その機能や内部構造をハードウェアとソフトウェア両面から調査し、製品に求められる要素を分析します。デザインは製品だけでなくサービスも対象とし、製品やサービスの背後に存在する代表的な技術と設計のプロセスを理解することで、実際の制作に活かすデザインエンジニアリングの方法論を身につけます。
類似度: 0.836, 
 ソリューション開発Ⅰ・Ⅱ 協力企業・団体から提供される実践的な課題をもとに、チームでソリューション（解決策）を考案し、1つのシステムを開発します。企画・開発したアイディアやプロトタイプに対して、企業から実際にフィードバックを受け、実社会で求められる問題解決能力を養います。ニーズに対してICTをどう活用するか、専門職人材として付加価値となる思考力を身につけます。
類似度: 0.834, 
 人工知能システム開発Ⅰ・Ⅱ 2年次前期、3年次前期の実習科目です。人工知能のシステム開発の専門知識と高度なスキルを身につけるため、代表的なアルゴリズムを応用してAIにおける要素技術を統合し、プロトタイプの作成、テスト、評価をチームで行います。システム開発に関しては、解決すべき課題、課題解決の機械学習アルゴリズムの選択、適用AIツールの選択、解決のためのシステム設計と実現方式、プログラム実装テスト、システム評価、問題解決度合いの最終評価など一連のシステム実装プロセスのプロトタイプ設計・試作（一部）を行い、実社会でのソリューションシステム開発をプロジェクトの進行として進める体験を行い、ノウハウを修得。AIを単に頭で理解するだけでなく、車の運転のように、全身で理解することができます。
類似度: 0.820, 
 東京国際工科専門職大学では、「プロフェッショナル」とは単に専門職に従事する人というだけではなく、その専門職における高い「職能」と「意識」を持ち、新しいイノベーションを生み出す人材を指しています。
類似度: 0.806, 
 深層学習 深層学習(Deep Learning) の主要技術であるニューラルネットワークの技術内容とその実現手法、効果に関して、講義・適用例の紹介・ツール活